### DISCUSSION TITLE
SSR support

### DISCUSSION THREAD

**Discussion by bdariusb** [TotallyInformation/node-red-contrib-uibuilder#228] [Ideas]
Hi, I couldn't figure out how to run multi page pre-rendered SvelteKit application, so I ran it as static generated page app with main index.html, because uibuilder requires that to be present at all times. However I would like it to run with server side rendering. So I assume one approach would involve generating a handler.js on build with svelte kit node adapter and passing it down to express js as written [here](https://kit.svelte.dev/docs/adapter-node#custom-server). Would that work with uibuilder?

**Comment by TotallyInformation** [comment]
Hi, yes, you should be able to run from a non-Node-RED server, the client allows for it. However, it isn't a well-tested approach so we might need to do some rapid tweaks. 😊

To use a 3rd-party server, you generally need to pass some parameters to the `uibuilder.start()` function that tells the client where the node-red server is and allows it to connect to the correct Socket.IO endpoint. These are documented. This lets you use a completely different web server to your Node-RED data/api server for example.

The alternative would be to use Node-RED/uibuilder to provide the full server. However, I've not really played with SvelteKit so I don't know what that would entail. But still, especially since uibuilder allows you to have an ExpressJS server on a separate port to Node-RED's server and with completely different settings, if you are planning to run your SSR from ExpressJS, it _should_ be possible to do with uibuilder. Untested of course but might be interesting to at least have a try-out to see if it is feasible as it might save you the effort of running a separate web server. Would be a nice contribution to the uibuilder community as well of course.

**Reply by bdariusb** [reply]
thanks for the answer. I managed to run a custom handler with custom express port. However there still a few problems to solve:
- if handler is an es module, it has to be configured in async manner. I use a dynamic import (import returns a promise):
```
                import(absolutePath)
                    .then(obj => this.app.use(obj.handler))
                    .catch(...)
```
- getting absolute/relative path for handler import. Not yet sure how to get absolute path to handler which is served from dist folder. Configuration happens in web.js
- bypass index.html check. Now it's a dummy file but, of course, there should be a nicer approach

**Reply by TotallyInformation** [reply]
Let me know if there is anything I can help with.

> bypass index.html check

Which check is that? I've been wondering whether I should still retain checks on index.html. It is good for people with less experience but could be somewhat limiting.
